source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,604 @@
+
+  // We must have enough fresh ids.
+  assert(fresh_ids.size() >= fresh_ids_needed && "Not enough fresh ids.");
+
+  // Update the module id bound
+  for (auto id : fresh_ids) {
+    fuzzerutil::UpdateModuleIdBound(ir_context, id);
+  }
+
+  // Create the block where the instruction is executed by splitting the
+  // original block.
+  auto execute_block = block->SplitBasicBlock(
+      ir_context, fresh_ids[0],
+      fuzzerutil::GetIteratorForInstruction(block, instruction));
+
+  // Create the merge block for the conditional that we are about to create by
+  // splitting execute_block (this will leave |instruction| as the only
+  // instruction in |execute_block|).
+  auto merge_block = execute_block->SplitBasicBlock(
+      ir_context, fresh_ids[1],
+      fuzzerutil::GetIteratorForInstruction(execute_block, next_instruction));
+
+  // Propagate the fact that the block is dead to the newly-created blocks.
+  if (transformation_context->GetFactManager()->BlockIsDead(block->id())) {
+    transformation_context->GetFactManager()->AddFactBlockIsDead(
+        execute_block->id());
+    transformation_context->GetFactManager()->AddFactBlockIsDead(
+        merge_block->id());
+  }
+
+  // Initially, consider the merge block as the alternative block to branch to
+  // if the instruction should not be executed.
+  auto alternative_block = merge_block;
+
+  // Add an unconditional branch from |execute_block| to |merge_block|.
+  execute_block->AddInstruction(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpBranch, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {merge_block->id()}}}));
+
+  // If the instruction requires 5 fresh ids, it means that it has a result id
+  // and its result needs to be used later on, and we need to:
+  // - add an additional block where a placeholder result is obtained by using
+  //   the OpUndef instruction
+  // - change the result id of the instruction to a fresh id
+  // - add an OpPhi instruction, which will have the original result id of the
+  //   instruction, in the merge block.
+  if (fresh_ids_needed == 5) {
+    // Create a new block using a fresh id for its label.
+    auto alternative_block_temp = MakeUnique<opt::BasicBlock>(
+        MakeUnique<opt::Instruction>(ir_context, SpvOpLabel, 0, fresh_ids[2],
+                                     opt::Instruction::OperandList{}));
+
+    // Keep the original result id of the instruction in a variable.
+    uint32_t original_result_id = instruction->result_id();
+
+    // Set the result id of the instruction to a fresh id.
+    instruction->SetResultId(fresh_ids[3]);
+
+    // Add an OpUndef instruction, with the same type as the original
+    // instruction and a fresh id, to the new block.
+    alternative_block_temp->AddInstruction(MakeUnique<opt::Instruction>(
+        ir_context, SpvOpUndef, instruction->type_id(), fresh_ids[4],
+        opt::Instruction::OperandList{}));
+
```

Comment:
Rather than creating an OpUndef here, I would prefer it if the transformation protobuf would contain an id of the right type that should be used.

The fuzzer pass can provide an irrelevant zero whenever possible, and an OpUndef when an irrelevant zero is not possible.

(We might create an OpUndef here as a last resort, in the case where we have to use overflow ids.)

Reply:
Could the transformation, instead, try to find a suitable zero constant? The fuzzer pass would ensure that it is in the module. This way, we minimise the amount of times that we need to use OpUndef (only for types that cannot be zero)

Reply:
The fuzzer pass will provide a suitable id. In the case where we have to use overflow ids, the transformation will try to find a suitable zero constant and it will use an OpUndef only if it fails to do so

Reply:
Done